src/libraries/Microsoft.Extensions.Logging.Abstractions/gen/LoggerMessageGenerator.Parser.cs

```diff
@@ -71,527 +71,528 @@ public IReadOnlyList<LoggerClass> GetLogClasses(IEnumerable<ClassDeclarationSynt
                 var eventIds = new HashSet<int>();
                 var eventNames = new HashSet<string>();
 
-                // we enumerate by syntax tree, to minimize the need to instantiate semantic models (since they're expensive)
-                foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax> group in classes.GroupBy(x => x.SyntaxTree))
+                // we only instantiate one semantic models per syntax tree since they're expensive
+                Dictionary<SyntaxTree, SemanticModel> semanticModels = classes
+                    .Select(c => c.SyntaxTree)
+                    .Distinct()
+                    .ToDictionary(st => st, st => _compilation.GetSemanticModel(st));
```

Comment:
This assumes that the `ClassDeclarationSyntax.SyntaxTree` returns a reference to the same instance for class declarations in the same syntax tree.